Show Appicon in notification

newhinton opened the pull request:
This adds the appicon to the notification. It adds a little touch to it and makes it look nicer imho. If the appicon is not fetched or some other error occured, the gotify icon is used

jmattheis commented on app/src/main/java/com/github/gotify/service/WebSocketService.java:
```diff
@@ -300,4 +336,44 @@ public void showNotificationGroup(long priority) {
+        return icon;
+    }
+
+    private Picasso makePicasso() {
```
Picasso is also created inside the message activity. Could you extract this method to com.github.gotify.Utils, to prevent code duplication?

newhinton replied:
@jmattheis The makePicasso-functions are not identical. Generally i would advise againts having a Utils class since they often fill up with stuff that should be somewhere else.

In this case, would it be fine with you if i create a chache-handler class that provides the appropriate methods?

jmattheis replied:
Yeah, feel free to make a class for this.

jmattheis commented on app/src/main/java/com/github/gotify/service/WebSocketService.java:
```diff
@@ -300,4 +336,44 @@ public void showNotificationGroup(long priority) {
+            return BitmapFactory.decodeResource(getResources(), R.drawable.gotify);
+        }
+
+        Bitmap icon;
+        RequestCreator request;
+        try {
+            request= cache.load(Utils.resolveAbsoluteUrl(settings.url() + "/", appIdMap.get(appid)));
+            icon = request.get();
+        } catch (IOException e) {
+            icon = BitmapFactory.decodeResource(getResources(), R.drawable.gotify);
+        }
+
+        return icon;
```
Return only and only declare variables where they needed
```suggestion
        try {
            RequestCreator request= cache.load(Utils.resolveAbsoluteUrl(settings.url() + "/", appIdMap.get(appid)));
            return request.get();
        } catch (IOException e) {
            Log.e("Could not load image", e);
            return BitmapFactory.decodeResource(getResources(), R.drawable.gotify);
        }
```

jmattheis commented on app/src/main/java/com/github/gotify/service/WebSocketService.java:
```diff
@@ -300,4 +336,44 @@ public void showNotificationGroup(long priority) {
+        try {
+            request= cache.load(Utils.resolveAbsoluteUrl(settings.url() + "/", appIdMap.get(appid)));
+            icon = request.get();
+        } catch (IOException e) {
```
Log the exception with `com.github.gotify.log.Log#e(java.lang.String, java.lang.Throwable)`

jmattheis commented on app/src/main/java/com/github/gotify/service/WebSocketService.java:
```diff
@@ -46,16 +62,17 @@
     private MissedMessageUtil missingMessageUtil;
 
+    private Picasso cache;
+    private Map<Integer, String> appIdMap;
```
This seems not thread save. Use ConcurrentHashMap to ensure thread safety Also the name can be more descriptive.
```suggestion
    private Map<Integer, String> appIdToAppImage = new ConcurrentHashMap<>();
```

jmattheis commented on app/src/main/java/com/github/gotify/service/WebSocketService.java:
```diff
@@ -46,16 +62,17 @@
     private AtomicInteger lastReceivedMessage = new AtomicInteger(NOT_LOADED);
     private MissedMessageUtil missingMessageUtil;
 
+    private Picasso cache;
```
While picasso does caching, it is not only a cache, so let's call it picasso.
```suggestion
    private Picasso picasso;
```

jmattheis commented on app/src/main/java/com/github/gotify/service/WebSocketService.java:
```diff
@@ -300,4 +336,44 @@ public void showNotificationGroup(long priority) {
+        return new Picasso.Builder(this).downloader(downloader).build();
+    }
+
+    private void updateAppIds(){
+        ApiClient client = ClientFactory.clientToken(settings.url(), settings.sslSettings(), settings.token());
+
+        List<Application> applications = null;
+        try {
+            applications = client.createService(ApplicationApi.class).getApps().execute().body();
+            appIdMap = MessageImageCombiner.appIdToImage(applications);
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
```
`#enqueue` can be used to asynchronously execute the http request. This removes the need for the `new Thread()..` inside `com.github.gotify.service.WebSocketService#startPushService` 
```suggestion
    private void updateAppIds(){
        ClientFactory.clientToken(settings.url(), settings.sslSettings(), settings.token())
                .createService(ApplicationApi.class)
                .getApps()
                .enqueue(Callback.call(
                                (apps) -> {
                                    appIdMap.clear();
                                    appIdMap.putAll(MessageImageCombiner.appIdToImage(apps));
                                },
                                (t) -> {
                                    appIdMap.clear();
                                }));
```
Using Callback.call automatically logs errors.

jmattheis commented on app/src/main/java/com/github/gotify/service/WebSocketService.java:
```diff
@@ -21,18 +23,32 @@
 import java.util.Map;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicInteger;
 
```
The travis-ci build currently fails because of formatting errors. Run `./gradlew spotlessApply` to fix them.
https://travis-ci.org/github/gotify/android/builds/677337457#L1868

jmattheis commented on app/src/main/java/com/github/gotify/messages/MessagesActivity.java:
```diff
@@ -120,7 +120,7 @@ public void onReceive(Context context, Intent intent) {
     private Integer selectAppIdOnDrawerClose = null;
 
-    int PICASSO_CACHE_SIZE = 50 * 1024 * 1024; // 50 MB
+    public static int PICASSO_CACHE_SIZE = 50 * 1024 * 1024; // 50 MB
```
This constant can be moved to Utils as well.

newhinton replied:
See the other comment regarding chache/picasso handler

newhinton commented:
I''ll just comment here, thanks for your feedback! I will change the code, i didn't now some of the stuff you pointed out.

jmattheis commented:
Inactivity.

newhinton commented:
@jmattheis Sorry for the inactivity, i have updated my pr. Would you still check it out? Should i make a new pr or can you reopen this one? 

I made the mistake and commited a file with my default fileheader wich contains my default GPLv3. I have changed that to MIT, is that enough?

jmattheis commented:
Yeah sure, you can re-open this PR. 

> I made the mistake and commited a file with my default fileheader wich contains my default GPLv3. I have changed that to MIT, is that enough?

Should be ok I guess.

newhinton commented:
I dont think that i can reopen it, since i am not a member of this project

jmattheis commented on app/src/main/java/com/github/gotify/messages/MessagesActivity.java:
```diff
@@ -120,9 +115,8 @@ public void onReceive(Context context, Intent intent) {
-    int PICASSO_CACHE_SIZE = 50 * 1024 * 1024; // 50 MB
-    private Cache picassoCache;
-    private Picasso picasso;
+    // private Picasso picasso;
```
```suggestion
```

jmattheis commented on app/src/main/java/com/github/gotify/PicassoHandler.java:
```diff
@@ -0,0 +1,118 @@
+    }
+
+    public void updateAppIds() {
+        Thread thread =
```
Apply this: https://github.com/gotify/android/pull/106#discussion_r412364887

jmattheis commented on app/src/main/java/com/github/gotify/PicassoHandler.java:
```diff
@@ -0,0 +1,118 @@
+    }
+
+    private void prepareCache() {
+        if (picassoCache == null) {
```
I think this is not needed, makePicasso should only be called once. The cache should be instantiated inside the constructor.

newhinton commented:
I have made changes, i hope i got everything

jmattheis commented:
Yeah, give me some days, I'm currently pretty busy.

jmattheis commented on app/src/main/java/com/github/gotify/picasso/PicassoHandler.java:
```diff
@@ -0,0 +1,103 @@
+import okhttp3.OkHttpClient;
+
+/**
+ * Copyright (C) 2020 Felix Nüsse Created on 01.06.20 - 20:03
```
Can we remove this? you still own the copyright regardless. I think this just bloats the source code, and is probably outdated from the start, because you copied some stuff from other files (like the makePicasso method).

newhinton replied:
Sure! I have it in my template, so its autogenerated. Will dump it later!

jmattheis replied:
Excellent, the rest of the code looks good, will merge it after your change.

jmattheis commented:
Could you rebase your branch?

newhinton commented:
this seems wrong
